pkg/credentialprovider/gcp/metadata.go

```diff
@@ -148,18 +166,73 @@ func (g *dockerConfigUrlKeyProvider) Provide() credentialprovider.DockerConfig {
 	return credentialprovider.DockerConfig{}
 }
 
+func runWithBackoff(url string, f func() ([]byte, error)) ([]byte, error) {
+	var backoff = time.Millisecond
+	const maxBackoff = time.Minute
+	for {
+		value, err := f()
+		if err == nil {
+			return value, nil
+		}
+		glog.V(2).Infof("failed to Get %q from gce metadata server: %v", url, err)
+		time.Sleep(backoff)
+		backoff = backoff * 2
+		if backoff > maxBackoff {
+			backoff = maxBackoff
+		}
+	}
+}
+
 // Enabled implements a special metadata-based check, which verifies the
 // storage scope is available on the GCE VM.
+// If running on a GCE VM, check if 'default' service account exists.
+// If it does not exist, assume that registry is not enabled.
+// If default service account exists, check if relevant scopes exist in the default service account.
+// The metadata service can become temporarily inaccesible. Hence all requests to the metadata
+// service will be retried until the metadata server returns a `200`.
+// It is expected that "http://metadata.google.internal./computeMetadata/v1/instance/service-accounts/" will return a `200`
+// and "http://metadata.google.internal./computeMetadata/v1/instance/service-accounts/default/scopes" will also return `200`.
+// More information on metadata service can be found here - https://cloud.google.com/compute/docs/storing-retrieving-metadata
 func (g *containerRegistryProvider) Enabled() bool {
-	value, err := credentialprovider.ReadUrl(metadataScopes+"?alt=json", g.Client, metadataHeader)
+	if !onGCEVM() {
+		return false
+	}
+	// Given that we are on GCE, we should keep retrying until the metadata server responds.
+	value, err := runWithBackoff(serviceAccounts, func() ([]byte, error) {
+		return credentialprovider.ReadUrl(serviceAccounts, g.Client, metadataHeader)
+	})
+	if err != nil {
+		glog.Errorf("failed to get list of service accounts: %v", err)
+		return false
+	}
+	// We expect the service account to return a list of account directories separated by newlines, e.g.,
```

Comment:
On a VM without a service account, the result is empty (with a 200 response). It looks like the following code will probably be ok with that, but please add it to the docs and test on a VM w/o a service account.